url89cafed56a_AlexDiru_esper_compiler_final_tgz-pJ8-compiler_ProgramJ8/src/compiler/EsperParser.java

```diff
@@ -0,0 +1,1997 @@
+            	case 1 :
+            	    // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:46:26: else_
+            	    {
+            	    pushFollow(FOLLOW_else__in_ifthenelse383);
+            	    else_29=else_();
+
+            	    state._fsp--;
+
+            	    adaptor.addChild(root_0, else_29.getTree());
+
+            	    }
+            	    break;
+
+            	default :
+            	    break loop6;
+                }
+            } while (true);
+
+
+            ENDIF30=(Token)match(input,ENDIF,FOLLOW_ENDIF_in_ifthenelse386); 
+            ENDIF30_tree = 
+            (Object)adaptor.create(ENDIF30)
+            ;
+            adaptor.addChild(root_0, ENDIF30_tree);
+
+
+            }
+
+            retval.stop = input.LT(-1);
+
+
+            retval.tree = (Object)adaptor.rulePostProcessing(root_0);
+            adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+
+        }
+        catch (RecognitionException re) {
+            reportError(re);
+            recover(input,re);
+    	retval.tree = (Object)adaptor.errorNode(input, retval.start, input.LT(-1), re);
+
+        }
+
+        finally {
+        	// do for sure before leaving
+        }
+        return retval;
+    }
+    // $ANTLR end "ifthenelse"
+
+
+    public static class if__return extends ParserRuleReturnScope {
+        Object tree;
+        public Object getTree() { return tree; }
+    };
+
+
+    // $ANTLR start "if_"
+    // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:47:1: if_ : IF ^ condition statements ;
+    public final EsperParser.if__return if_() throws RecognitionException {
+        EsperParser.if__return retval = new EsperParser.if__return();
+        retval.start = input.LT(1);
+
+
+        Object root_0 = null;
+
+        Token IF31=null;
+        EsperParser.condition_return condition32 =null;
+
+        EsperParser.statements_return statements33 =null;
+
+
+        Object IF31_tree=null;
+
+        try {
+            // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:47:5: ( IF ^ condition statements )
+            // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:47:7: IF ^ condition statements
+            {
+            root_0 = (Object)adaptor.nil();
+
+
+            IF31=(Token)match(input,IF,FOLLOW_IF_in_if_394); 
+            IF31_tree = 
+            (Object)adaptor.create(IF31)
+            ;
+            root_0 = (Object)adaptor.becomeRoot(IF31_tree, root_0);
+
+
+            pushFollow(FOLLOW_condition_in_if_397);
+            condition32=condition();
+
+            state._fsp--;
+
+            adaptor.addChild(root_0, condition32.getTree());
+
+            pushFollow(FOLLOW_statements_in_if_399);
+            statements33=statements();
+
+            state._fsp--;
+
+            adaptor.addChild(root_0, statements33.getTree());
+
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1119:1144, 1889:1914.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.